modules/juce_graphics/effects/juce_StackShadowEffect.cpp

```diff
@@ -0,0 +1,329 @@
+    22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22,
+    22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 23,
+    23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23,
+    23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23,
+    23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23,
+    23, 23, 23, 23, 23, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24,
+    24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24,
+    24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24,
+    24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24,
+    24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24
+};
+
+// The Stack Blur Algorithm was invented by Mario Klingemann,
+// mario@quasimondo.com and described here:
+// http://incubator.quasimondo.com/processing/fast_blur_deluxe.php
+
+// Stackblur algorithm by Mario Klingemann
+// Details here:
+// http://www.quasimondo.com/StackBlurForCanvas/StackBlurDemo.html
+// C++ implemenation base from:
+// https://gist.github.com/benjamin9999/3809142
+// http://www.antigrain.com/__code/include/agg_blur.h.html
+static void applyStackBlur (Image& img, unsigned int radius)
+{
+    const unsigned int w = (unsigned int)img.getWidth();
+    const unsigned int h = (unsigned int)img.getHeight();
+
+    Image::BitmapData data (img, Image::BitmapData::readWrite);
+
+    radius = jlimit (2u, 254u, radius);
+
+    unsigned char stack[(254 * 2 + 1) * 1];
+
+    unsigned int x, y, xp, yp, i, sp, stack_start;
+
+    unsigned char* stack_ptr = nullptr;
+    unsigned char* src_ptr = nullptr;
+    unsigned char* dst_ptr = nullptr;
+
+    unsigned long sum, sum_in, sum_out;
+
+    unsigned int wm = w - 1;
+    unsigned int hm = h - 1;
+    unsigned int w1 = (unsigned int) data.lineStride;
+    unsigned int div = (unsigned int) (radius * 2) + 1;
+    unsigned int mul_sum = stackblur_mul[radius];
+    unsigned char shr_sum = stackblur_shr[radius];
+
+    for (y = 0; y < h; ++y)
+    {
+        sum = sum_in = sum_out = 0;
+
+        src_ptr = data.getLinePointer (int (y));
+
+        for (i = 0; i <= radius; ++i)
+        {
+            stack_ptr    = &stack[i];
+            stack_ptr[0] = src_ptr[0];
+            sum += src_ptr[0] * (i + 1);
+            sum_out += src_ptr[0];
+        }
+
+        for (i = 1; i <= radius; ++i)
+        {
+            if (i <= wm)
+                src_ptr += 1;
+
+            stack_ptr = &stack[1 * (i + radius)];
+            stack_ptr[0] = src_ptr[0];
+            sum += src_ptr[0] * (radius + 1 - i);
+            sum_in += src_ptr[0];
+        }
+
+        sp = radius;
+        xp = radius;
+        if (xp > wm)
+            xp = wm;
+
+        src_ptr = data.getLinePointer (int (y)) + (unsigned int)data.pixelStride * xp;
+        dst_ptr = data.getLinePointer (int (y));
+
+        for (x = 0; x < w; ++x)
+        {
+            dst_ptr[0] = (unsigned char)((sum * mul_sum) >> shr_sum);
```

Comment:
Could reduce the size of the nested loops, and aid readability by encapsulating them in free functions, something like:

```
for (auto y = 0; y < h; y++)
{
    // ...

    for (auto x = 0; x < w; x++)
    {
        blurHorizontally (x, y);
    }
}
```